src/dbnode/storage/namespace/schema.go

```diff
@@ -0,0 +1,226 @@
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package namespace
+
+import (
+	"bytes"
+	"compress/gzip"
+	"errors"
+	"io/ioutil"
+	"github.com/golang/protobuf/proto"
+	dpb "github.com/golang/protobuf/protoc-gen-go/descriptor"
+	"github.com/jhump/protoreflect/desc"
+	nsproto "github.com/m3db/m3/src/dbnode/generated/proto/namespace"
+	xerrors "github.com/m3db/m3x/errors"
+	"strings"
+)
+
+var (
+	errInvalidSchema        = errors.New("invalid schema definition")
+	errSchemaNotFound       = errors.New("schema is not found")
+	errSchemaRegistryEmpty  = errors.New("schema registry is empty")
+	errInvalidSchemaOptions = errors.New("invalid schema options")
+)
+
+type schemaDescr struct {
+	deployId string
+	md       *desc.MessageDescriptor
+}
+
+func (s *schemaDescr) DeployId() string {
+	return s.deployId
+}
+
+func (s *schemaDescr) Equal(o SchemaDescr) bool {
+	if s == nil && o == nil {
+		return true
+	}
+	if s != nil && o == nil || s == nil && o != nil {
+		return false
+	}
+	if _, ok := o.(*schemaDescr); !ok {
+		return false
+	}
+	return s.DeployId() == o.DeployId()
+}
+
+func (s *schemaDescr) Get() *desc.MessageDescriptor {
+	return s.md
+}
+
+func (s *schemaDescr) String() string {
+	if s.md == nil {
+		return ""
+	}
+	return s.md.String()
+}
+
+type schemaRegistry struct {
+	options  *nsproto.SchemaOptions
+	latestId string
+	// a map of schema version to schema descriptor.
+	versions map[string]*schemaDescr
+}
+
+func (sr *schemaRegistry) Equal(o SchemaRegistry) bool {
+	var osr *schemaRegistry
+	var ok bool
+	if sr == nil && o == nil {
+		return true
+	}
+	if sr != nil && o == nil || sr == nil && o != nil {
+		return false
+	}
+
+	if osr, ok = o.(*schemaRegistry); !ok {
+		return false
+	}
+	// compare latest version
+	if sr.latestId != osr.latestId {
+		return false
+	}
+
+	// compare version map
+	if len(sr.versions) != len(osr.versions) {
+		return false
+	}
+	for v, sd := range sr.versions {
+		osd, ok := osr.versions[v]
+		if !ok {
+			return false
+		}
+		if !sd.Equal(osd) {
+			return false
+		}
+	}
+
+	return true
+}
+
+func (sr *schemaRegistry) Get(id string) (SchemaDescr, error) {
```

Comment:
Might be a little cleaner to just have this return `(SchemaDescr, bool)` and let the caller construct an error if that is truly an error scenario otherwise they have to do some kind of error comparison to see if its `errSchemaNotFound` which isnt even exported